### PR TITLE
Support YAML validation through JSON schema

### PR DESCRIPTION
**Pull request by robrechtme**
## Proposed Changes

JSON schema can be used for [validating YAML files](https://json-schema-everywhere.github.io/yaml). This MR serves as a first draft to pour the Maestro docs in a JSON schema, which can be used for validating the syntax of Maestro flows.

It is far from complete, and I need some help on this because it is a lot of work 🙏 

If finished, the schema can be submitted to the [SchemaStore](https://github.com/SchemaStore/schemastore) so validation can be done automatically based on a file name.

## Testing
1. Install the VSCode [YAML](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) extension 
2. Copy the schema.json file into your project, next to your Maestro flows
3. Configure the schema in `.vscode/settings.json`

```
"yaml.schemas": {
    "./path/to/schema.json": "path/to/flows/**/*.yaml",
  }
```

## Issues Fixed

Fixes https://github.com/mobile-dev-inc/maestro/issues/467

### CODE REVIEW

File: schema/schema.json [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        "title": "Assert whether an element is not visible.",
+        "type": ["string", "object"],
+        "additionalProperties": false,
+        "properties": {
+          "text": { "type": "string" },
+          "id": { "type": "string" },
+          "enabled": { "type": "boolean" },
+          "checked": { "type": "boolean" },
+          "focused": { "type": "boolean" },
+          "selected": { "type": "boolean" }
```
Comment (simon-gilmurray) [inline schema/schema.json:19]:
```
"selected": { "type": "boolean" },
"below": { "$ref": "#/$defs/Element" },
"leftOf": { "$ref": "#/$defs/Element" },
"rightOf": { "$ref": "#/$defs/Element" }
```

File: schema/schema.json [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        "type": ["string", "object"],
+        "additionalProperties": false,
+        "properties": {
+          "text": { "type": "string" },
+          "id": { "type": "string" },
+          "index": { "type": "integer" },
+          "enabled": { "type": "boolean" },
+          "checked": { "type": "boolean" },
+          "focused": { "type": "boolean" },
+          "selected": { "type": "boolean" }
```
Comment (simon-gilmurray) [inline schema/schema.json:33]:
```
"selected": { "type": "boolean" },
"below": { "$ref": "#/$defs/Element" },
"leftOf": { "$ref": "#/$defs/Element" },
"rightOf": { "$ref": "#/$defs/Element" }
```

File: schema/schema.json [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+          "timeout": { "type": "integer" },
+          "speed": { "type": "integer" },
+          "visibilityPercentage": { "type": "integer" }
+        }
+      },
+      "swipe": {
+        "title": "Perform a swipe gesture.",
+        "type": "object",
+        "additionalProperties": false,
+        "properties": {
```
Comment (WookieFPV) [inline schema/schema.json:163]: There are a few other ways how swipes can be done:
https://maestro.mobile.dev/api-reference/commands/swipe

`"start":{
              "type": "string"
            },
            "end": {
              "type": "string"
            }`
            
         I don't know JSON schema that well but that works for me

**Comment by filipef101** [comment]
Is this still missing properties?

**Comment by simon-gilmurray** [comment]
> Is this still missing properties?

@filipef101 Yes, this was created 3 months ago, was missing some then and Maestro has moved on a bit since then.
I believe the team were going to look in to an alternative to this, but not sure on the progress.

**Comment by riglar** [comment]
We've stood up one here: [https://www.moropo.com/maestro-yaml-validator](https://www.moropo.com/maestro-yaml-validator)

Hopefully useful to someone until the official one is available.

**Comment by robrechtme** [comment]
@riglar your link doesn't seem to work 🤔

**Comment by filipef101** [comment]
@robrechtme Can you allow me to update this PR or should I raise a new one updating this?

**Comment by robrechtme** [comment]
@filipef101 I resolved the remarks so I guess it's ready to merge?

**Comment by filipef101** [comment]
@robrechtme Ah okay, only saw 1 commit

**Comment by robrechtme** [comment]
I amended my changes to the first commit to keep the git history clean

**Comment by riglar** [comment]
> @riglar your link doesn't seem to work 🤔

Apologies - we did have some downtime to make some improvements - should be working now

**Comment by robrechtme** [comment]
@riglar awesome! Are you willing to share the JSON schema that’s used under the hood, or is this not using a JSON schema?

**Comment by filipef101** [comment]
missing evalScript

**Comment by axelniklasson** [comment]
Hi there @robrechtme 👋 Axel from mobile.dev here. Thanks for raising this PR! This has grown quite stale, so if you (or anyone else) are still up for updating it to be inline with the latest API spec I'd be happy to help this get merged.

**Comment by robrechtme** [comment]
@axelniklasson Hi, thanks for looking at this. I'd propose to merge this as a first step, so it's easier for anyone to contribute on this any further?

**Comment by axelniklasson** [comment]
> @axelniklasson Hi, thanks for looking at this. I'd propose to merge this as a first step, so it's easier for anyone to contribute on this any further?

I'll try to find some time to do an inventory of what is there / missing throughout this week so I know what's not there before landing. I'm hesitant to having partial support in this schema and the team is heads down on other things at the moment so I worry that this will just stay with partial support. That said, I will try to find some time and then I can possible help out getting this across the line :)

**Comment by axelniklasson** [comment]
Following up here -- week has been filled with other things so I won't be able to dig in here. If you're open to updating this to cover the current API spec I'll be happy to merge it but otherwise I'll have to close this one out as we don't want to merge something that covers part of the spec.

**Comment by axelniklasson** [comment]
Closing this one out due to inactivity -- @robrechtme let me know if you plan on tackling this and we can take it from there! I appreciate the work that went into this and hope that you understand that we're unable to prioritize this right now and want to refrain from merging something incomplete. If you end up shipping this somewhere else, let us know and we'll be happy to link to it from our [Community contributions page](https://maestro.mobile.dev/community/contributions) 🙂

**Comment by riglar** [comment]
@axelniklasson - thanks for the update and offering to link our version of the validator from the community page.

The link is here: https://www.moropo.com/maestro-yaml-validator

And we've been keeping it up to date as new Maestro versions are released.

**Comment by bartekpacia** [comment]
Hey @riglar and @robrechtme 👋 Bartek from mobile.dev here:)

I think there are some great ideas in this issue. I'll reopen it so community can chime in & contribute. Maybe we could:

1. contribute a schema to SchemaStore, so validation would work out-of-the-box for Maestro flows. We could say that Maestro flows are:
  - YAML files
  - with names ending with `_test.yaml` or _flow.yaml`
  - inside the `.maestro` directory (recursively)
2. show documentation for command/command's arguments on mouse hover in VSCode/other IDEs. (Maybe it's already handled by bullet point 1)?

It'd be really great if the schema could be auto-generated from [Commands.kt](https://github.com/mobile-dev-inc/maestro/blob/9b68a2b2743d671f7584a370f2f9f3887c34a40c/maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt), to reduce maintenance required.

**Comment by bartekpacia** [comment]
Oops, I meant I will create a new issue, sorry

**Pull request closed without merging**